feat(CodeEditor): Changing schemas based on the content

mmelko opened the pull request:
This PR introduces EntitySchemaConfig which is meant to be to replace SchemasStore. Schemas should be possible to configure within the entities-context. Also this PR introduces `currentEntity`field which is used for holding the current `EntityType`.  After change of the `currentEntity` proper schema should be also reloaded in the CodeEditor.

related to #136 #135

lordrip commented on packages/ui/src/components/SourceCode/SourceCode.tsx:
```diff
@@ -2,10 +2,11 @@ import { CodeEditor, CodeEditorProps, Language } from '@patternfly/react-code-ed
 import { SyncButton } from './SyncButton';
 import './workers/enable-workers';
+import { useEntities } from '../../hooks';
+import { entitySchemaConfig } from '../../models/camel-entities/EntitySchemaConfig.ts';
```
```suggestion
import { entitySchemaConfig } from '../../models/camel-entities/EntitySchemaConfig';
```

lordrip commented on packages/ui/src/hooks/entities.ts:
```diff
@@ -108,3 +123,15 @@ function getEntity(rawEntity: unknown): BaseCamelEntity | BaseVisualCamelEntity
+  }
+  if (isPipe(entity)) return EntityType.Pipe;
+  return EntityType.Pipe;
+}
```
I think that in this case, we need to set the `Camel YAML DSL` schema as default, and only change the schema for:

* KameletBindings
* Pipes
* Kamelet

So then the code should read more or less like:

```
  if (isKameletBinding(entity)) return EntityType.KameletBinding;

  if (isKamelet(entity)) return EntityType.Kamelet;

  if (isPipe(entity)) return EntityType.Pipe;

  if (isIntegration(entity)) return EntityType.Integration;

  // After reaching this point, it should be enough to set the `Camel YAML DSL` schema
  return EntityType.Route;
```

mmelko replied:
done

lordrip commented on packages/ui/src/models/camel-entities/EntitySchemaConfig.ts:
```diff
@@ -0,0 +1,86 @@
+  [EntityType.KameletBinding]: IEntitySchema;
+  [EntityType.Integration]: IEntitySchema;
+  [EntityType.Rest]: IEntitySchema;
+  [EntityType.RestConfiguration]: IEntitySchema;
```
I think that for DSL purposes, we should only have the ones that affects the Source Code, meaning that in this interface it should be:

* Kamelet
* KameletBinding
* Integration
* Pipe
* Route

mmelko replied:
Thanks. I agree I was just following what we have before.

lordrip commented on packages/ui/src/providers/catalog-schema-loader.provider.tsx:
```diff
@@ -2,6 +2,7 @@ import { FunctionComponent, PropsWithChildren, createContext, useEffect, useStat
 import { CamelSchemasProcessor, DEFAULT_CATALOG_PATH } from '../camel-utils';
 import { CamelCatalogIndex, CatalogEntry, CatalogKind, ComponentsCatalog, Schema } from '../models';
 import { useCatalogStore, useSchemasStore } from '../store';
+import { entitySchemaConfig } from '../models/camel-entities/EntitySchemaConfig.ts';
```
```suggestion
import { entitySchemaConfig } from '../models/camel-entities/EntitySchemaConfig';
```

lordrip commented on packages/ui/src/models/camel-entities/EntitySchemaConfig.ts:
```diff
@@ -0,0 +1,86 @@
+    },
+  };
+
+  setSchema(schema: Schema) {
```
The schemas should have name already, so I guess this can be be simplified

mmelko commented:
I updated the PR, incorporated your suggestion and:
- new enum `SourceSchemaType` ..  as we discussed
- renamed `currentEntity` to `currentSchemaType`
- renamed  `EntitySchemaConfig` to `SourceSchemaConfig` 
-  add `isKamelet` and `isIntegration` with stubs and tests to being able to assign the schema properly 

If the names are off .. I can change them :)

lordrip commented on packages/ui/src/hooks/entities.ts:
```diff
@@ -21,17 +26,18 @@ export const useEntities = (): EntitiesContextResult => {
       const entitiesHolder = rawEntities.reduce(
         (acc, rawEntity) => {
           const entity = getEntity(rawEntity);
+          setCurrentSchemaType(getEntityType(rawEntity));
```
[important] I think we don't need to set the schema inside of this iteration, as it leads to setting many times the same value, potentially firing subsequent updates.

I guess that it will be enough to inspect only the first element of the `rawEntities` array to determine which schema to apply just once.

lordrip commented on packages/ui/src/models/camel-entities/SourceSchemaConfig.ts:
```diff
@@ -0,0 +1,52 @@
+    },
+  };
+
+  setSchema(name: string, schema: Schema) {
+    const type: SourceSchemaType = SourceSchemaType[name];
+    if (name === 'camelYamlDsl') {
+      this.config[SourceSchemaType.Route].schema = schema;
+    }
+    if (type) {
+      this.config[type].schema = schema;
+    }
+  }
```
I saw the `TS` issue that you mentioned and I made a few changes to the method:

```typescript
  setSchema(name: string | SourceSchemaType, schema: Schema) {
    if (name === 'camelYamlDsl') {
      this.config[SourceSchemaType.Route].schema = schema;
      return;
    }

    if (isEnumType(name, SourceSchemaType)) {
      const type: SourceSchemaType = SourceSchemaType[name];
      this.config[type].schema = schema;
    }
  }
```

The `isEnumType` function is a special type of function, a Typeguard function, now called [`Type predicate`](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates)

It's defined like the following:
```typescript
function isEnumType<T extends object>(type: unknown, enumObject: T): type is keyof T {
  return typeof type === 'string' && (type as keyof typeof enumObject) in enumObject;
}
```

You could use that function and place it in a utils folder, maybe alongside of `packages/ui/src/utils/is-enum-type.ts`

mmelko replied:
Many thanks!

lordrip reviewed (approved):
It looks great @mmelko , thanks for putting this together, I really like updating the schema upon writing, very cool.
